Refactor world tile related codes (WIP)

badewen opened the pull request:
Let me know if there are anything that needs changes. (This is not backward compatible as it changes some entries in the database)
Here is roughly the changes (20/05/2025)
```
added realloc, grow, shrink on ExtendBuffer
renamed Block class name to TileData
renamed id_user to userID in PeerData
changed userID in PeerData to be strictly number type only
added hasTilePermission method to World class
changed how World class loads its Tile. It uses Tile.from to construct the Tile class according to the item ActionType. Example, invoking Tile.from with TileData containing the ID of door, will return DoorTile.ts.
when saving World tile data to database, it will save Tile.data instead of the Tile itself.
added TileFlag into TileData
added enum LockPermission (Constant.ts) to be used in LockedBlocked to simplify permission checking. Also removed  onlyAllowBuild field in favor of the aforementioned field.
added default permission for locks in Constant.ts
```

EDIT: To anyone trying to test this branch, just apply this commit f1f4a8e5b1eb6b8f06d5eaa0a0b81d019385813e

badewen commented:
Yup that was intentional. Placing white door without giving it any tile
extra causes it to act weird, but the whole code works as intended.

On Wed, May 21, 2025, 9:52 PM Hendra Gunawan ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/world/generation/Default.ts
> <https://github.com/StileDevs/GrowServer/pull/137#discussion_r2100498090>:
>
> >          };
>
> -        if (block.y === Y_START_DIRT - 1 && block.x === mainDoorPosition) {
> +        if (false) {
>
> Seems typescript complaining this code
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/StileDevs/GrowServer/pull/137#pullrequestreview-2858046201>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ATPUARBYFGAKFMDLPXO7LOD27SHKDAVCNFSM6AAAAAB5Q5LN66VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDQNJYGA2DMMRQGE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

JadlionHD commented:
Ok, let me know when this ready to review

badewen commented:
It is kinda ready for review. Any thoughts on the design pattern? If not,
then i'm going to continue with this design pattern.

On Wed, May 21, 2025, 11:48 PM Hendra Gunawan ***@***.***>
wrote:

> *JadlionHD* left a comment (StileDevs/GrowServer#137)
> <https://github.com/StileDevs/GrowServer/pull/137#issuecomment-2898610447>
>
> Ok, let me know when this ready to review
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/StileDevs/GrowServer/pull/137#issuecomment-2898610447>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ATPUARACBDD3N4D5UOLA2PL27SU6PAVCNFSM6AAAAAB5Q5LN66VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQOJYGYYTANBUG4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

JadlionHD commented on src/core/Peer.ts:
```diff
@@ -98,6 +98,9 @@ export class Peer extends OldPeer<PeerData> {
     });
   }
 
+  // i kinda hate how this is called everytime just to send packets to a specific world. 
```
It does reduce logical CPU usage when checking peers in worlds.
But it would consume alot of memory when storing Peers object directly into cache.

JadlionHD commented on src/network/dialogs/AreaLockEdit.ts:
```diff
@@ -1,93 +1,93 @@
-import { World } from "../../core/World";
-import { Tile } from "../../world/Tile";
-import { ItemDefinition } from "growtopia.js";
+// import { type NonEmptyObject } from "type-fest";
```
Any reason commenting this?

JadlionHD replied:
I meant the whole file

badewen replied:
https://github.com/StileDevs/GrowServer/pull/137#issuecomment-2899766778

JadlionHD commented on src/network/dialogs/DoorEdit.ts:
```diff
@@ -1,57 +1,57 @@
-import { World } from "../../core/World";
-import { Tile } from "../../world/Tile";
-import { ItemDefinition } from "growtopia.js";
+// import { type NonEmptyObject } from "type-fest";
```
This file also

badewen replied:
https://github.com/StileDevs/GrowServer/pull/137#issuecomment-2899766778

JadlionHD commented on src/network/dialogs/SignEdit.ts:
```diff
@@ -1,53 +1,53 @@
-import { World } from "../../core/World";
-import { Tile } from "../../world/Tile";
-import { ItemDefinition } from "growtopia.js";
+// import { type NonEmptyObject } from "type-fest";
```
This

badewen replied:
https://github.com/StileDevs/GrowServer/pull/137#issuecomment-2899766778

badewen commented:
Make sense, but it is getting cumbersome to write it each time when
broadcasting into peers in the same world. Will probably wrap it in
World.broadcast method and later, figure out the best way to do it

On Thu, May 22, 2025, 9:48 AM Hendra Gunawan ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/core/Peer.ts
> <https://github.com/StileDevs/GrowServer/pull/137#discussion_r2101517811>:
>
> > @@ -98,6 +98,9 @@ export class Peer extends OldPeer<PeerData> {
>      });
>    }
>
> +  // i kinda hate how this is called everytime just to send packets to a specific world.
>
> It does reduce logical CPU usage when checking peers in worlds.
> But it would consume alot of memory when storing Peers object directly
> into cache.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/StileDevs/GrowServer/pull/137#pullrequestreview-2859617277>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ATPUAREDUCWK6G3A47PR3GL27U3HHAVCNFSM6AAAAAB5Q5LN66VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDQNJZGYYTOMRXG4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

JadlionHD reviewed (commented):
Also what will you do with the rest of tiles files?

badewen commented:
I commented it because i havent implemented any locks yet, also to get it
to work minimally. Currently i have plan
 to store list of owned tile index into the lock owner data and maybe
rework how it uninitialize all of the owned tile (havent looked into the
codes though). Will discuss it in discord if i get to that point.

On Thu, May 22, 2025, 9:53 AM Hendra Gunawan ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/network/dialogs/AreaLockEdit.ts
> <https://github.com/StileDevs/GrowServer/pull/137#discussion_r2101521673>:
>
> > @@ -1,93 +1,93 @@
> -import { type NonEmptyObject } from "type-fest";
> -import { Base } from "../../core/Base";
> -import { Peer } from "../../core/Peer";
> -import { LOCKS } from "../../Constants";
> -import { Block, Lock } from "../../types";
> -import { Floodfill } from "../../utils/FloodFill";
> -import { World } from "../../core/World";
> -import { Tile } from "../../world/Tile";
> -import { ItemDefinition } from "growtopia.js";
> +// import { type NonEmptyObject } from "type-fest";
>
> I meant the whole file
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/StileDevs/GrowServer/pull/137#discussion_r2101521673>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ATPUARC6A7N3VTPOL63NK2D27U3YZAVCNFSM6AAAAAB5Q5LN66VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDQNJZGYZDEMJUGM>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

badewen commented:
I commented all of the files because i havent worked on it yet. Just to get it minimally working.

JadlionHD commented:
Looks good so far, unless with that `every` function, maybe stick with those for now. since storing it into cache would be ram expensive apparently.

JadlionHD commented:
> Make sense, but it is getting cumbersome to write it each time when broadcasting into peers in the same world. Will probably wrap it in World.broadcast method and later, figure out the best way to do it


Oh thats good thing todo with

badewen commented:
> > Make sense, but it is getting cumbersome to write it each time when broadcasting into peers in the same world. Will probably wrap it in World.broadcast method and later, figure out the best way to do it
> 
> Oh thats good thing todo with

Okay then, will refrain from storing the peers to the world for now.

badewen commented:
```
- implement WeatherTile.ts, SeedTile.ts, LockTile.ts
- when destroying tree, gives 10% chance of returning the seed. Also applies when harvesting tree 
- fixed and reworked some permission checks
- added some placeholder methods in Tile.ts (onPlayerPass, onStep)
- added ownedTiles field in LockedBlocked class
- simplify hasTilePermission implementation
- implement sendTextBubble, sendSFX, sendOnPlayPositioned, sendConsoleMessage in Peer.ts
- moved playing use_lock.wav from FloodFill.apply to LockTile.ts
- fix SeedTile.ts
- Compress SEND ITEM DATABASE
- Added addPlayerPicker method to DialogBuilder
- Added getByUID method to PlayerDB
- Added getPeerByNetID to World.ts
- Partially implemented access in locks. (To be implemented: Wrench your self to accept, world lock key) 
- Added RevokeLockAccess on the dialog handler to handle self revoke access.
- Added new min_level field into WorldDB and WorldData.
- Paritally working World Lock
```

badewen commented:
- moved World lock related data into WorldLockData
- cleaned up some unused interface in World.d.ts
- refactored LockedBlocked into 2 classes (LockChild and LockParent)
- added getOwnerUID to world.ts
- made entering and leaving world text bubble stacked
- removed name field from HeartMonitor
- added return value to on* functions in Tile.ts
- implement HeartMonitor.ts
- added heart_monitors to the Player database and interface to track placed HeartMonitor
- Cherry pick 177346a012b62820ed6bfc3817dd436b327ac947, 97c25f0f29e3c28a705fe2b0f1ee0ec980f8d886
- Ported ItemDefinition from growtopia.js to grow-items
- Added tielUpdateMultiple to index.ts in Tiles folder.
- Fixed entering world from world list with player in it.

JadlionHD commented:
Alright I'll merge this, since I making it into monorepo (I will make the PR asap)

badewen commented:
> Alright I'll merge this, since I making it into monorepo (I will make the PR asap)

alright good to hear 😄
